### DISCUSSION TITLE
Where in the source code would this problem originate from?

### DISCUSSION THREAD

**Discussion by TimRegeant** [NationalSecurityAgency/ghidra#6241] [Q&A]
For 68000, the cmpi.b instuction is not disassembled correctly. You can see below the immediate value shows in the Listing window as a word instead of byte and further in the second occurance it references a DAT address label instead of showing an immediate value.

Example:
```
  0001a9e0 0c 00 00 ff    cmpi.b    #0xFFFF,D0b
  0001a9e4 67 00 00 88    beq.w     LAB_0001aa6e
  0001a9e8 0c 00 00 fe    cmpi.b    #PTR_DAT_0000fffe,D0b

``` 
I would like to look in the source to try to fix this, but not sure where this problem would arise.  

I am thinking it might be in the 68000.sinc file...is that correct?  Thank you.

**Comment by astrelsky** [comment]
Should be https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Processors/68000/data/languages/68000.sinc#L1391

Note: If you right click on an instruction in the listing and select instruction info, it will tell you the exact file and line the constructors for the instruction and its tables are defined.

**Comment by pjsoberoi** [comment]
Thanks for the link. That being said why is it two-bytes and not one byte like the user mentioned? The constructors look correct to me. d8 is defined as:

```
  d8     = (0,7) signed
```
So 8-bits but signed. The UI is sign-extending to 16-bits. How do we disable that? I might have time to look at it tomorrow.

**Reply by jayoung99** [reply]
It seems to be only happening with the mode 0 form of the instruction where a data register is used.

And this mode 0 form isn't broken on the 'andi.b' instruction.  It uses different semantics and constraints:

`:andi.b const8,e2b		is opbig=2 & op67=0 & $(DAT_ALTER_ADDR_MODES); const8; e2b		[ savmod2=savmod1; regtsan=regtfan; ]	{ and(const8, e2b); }
`

Might be a clue to why this is happening.  Contrast with 'cmpi.b''s constructor:

`:cmpi.b "#"^d8,e2b			is opbig=12 & op67=0 & savmod1 & regtfan & $(DAT_ALTER_ADDR_MODES); d8; e2b	[ savmod2=savmod1; regtsan=regtfan; ]	{ o2:1=e2b; subflags(o2,d8); local tmp =o2-d8; resflags(tmp); }
`

`

**Reply by TimRegeant** [reply]
If it can help there is this instruction layout.

```
     00000110  00 000 000 00000000 00000000
          |     |  |   |      |        |
cmpi -----'     |  |   |      |        |
size -----------'  |   |      |        |
mode --------------'   |      |        |
register --------------'      |        |
16-bit data ------------------+--------'
8-bit data ----------------------------'

```

**Comment by jayoung99** [comment]
Have developed a fix for this issue.

The issue revolved around the d8 field being defined in the extword token which is a 16 bit token.

So for some underlying reason, since d8 was defined as signed and was contained in a 16 bit token, Ghidra was sign extending to all 16 bits.

The work around I came up with was to define two new 8 bit tokens to use in place of the 16 bit extword token for the cmpi.b instruction:

```
define token extwordb1 (8)
  ignore     = (0,7)
;
define token extwordb2 (8)
  d8b    = (0,7) signed
;
```

Then I changed the cmpi.b entry in the root table to use these 2 byte tokens instead of the 16 bit extword token.

```
#:cmpi.b "#"^d8,e2b		is opbig=12 & op67=0 & savmod1 & regtfan & $(DAT_ALTER_ADDR_MODES); d8; e2b	[ savmod2=savmod1; regtsan=regtfan; ]	{ o2:1=e2b; subflags(o2,d8); local tmp =o2-d8; resflags(tmp); }
:cmpi.b "#"^d8b,e2b		is opbig=12 & op67=0 & savmod1 & regtfan & $(DAT_ALTER_ADDR_MODES); ignore; d8b; e2b	[ savmod2=savmod1; regtsan=regtfan; ]	{ o2:1=e2b; subflags(o2,d8b); local tmp =o2-d8b; resflags(tmp); }
```

This results in Ghidra only sign extending to the 8 bit value in the lower byte which displays the correct value in the Listing.
